File: app/services/knight_tour_game.rb
class KnightTourGame
  class IllegalMoveError < StandardError; end

  attr_reader :tour

  def initialize(tour:)
    @tour = tour
  end

  def current_square
    last = tour.moves.order(:position).last
    last && Square.from_notation(last.square)
  end

  def visited?(square)
    tour.moves.exists?(square: square.notation)
  end

  def legal_moves_from
    return Square.all if tour.moves.none?
    MoveFinder.new(square: current_square).legal_moves.reject { |square| visited?(square) }
  end

  def visit!(square)
    raise IllegalMoveError, "#{square.notation} is not legal" unless legal_moves_from.include?(square)
    tour.moves.create!(square: square.notation, position: next_position)
  end

  def undo!
    tour.moves.order(:position).last&.destroy
  end

  def visited_count
    tour.moves.count
  end

  def won?
    visited_count == 64
  end

  def stuck?
    visited_count.positive? && !won? && legal_moves_from.empty?
  end

  private

  def next_position
    (tour.moves.maximum(:position) || 0) + 1
  end
end

File: app/javascript/game/knight_tour_game.js
import { Square } from "#game/square"
import { MoveFinder } from "#game/move_finder"

export class IllegalMoveError extends Error {}

export class KnightTourGame {
  constructor() {
    this.moves = []
    this.redoStack = []
  }

  get currentSquare() { return this.moves.length ? this.moves[this.moves.length - 1] : null }
  get lastMove() { return this.currentSquare }

  visited(square) { return this.moves.some(m => m.equals(square)) }

  get legalMovesFrom() {
    if (this.moves.length === 0) return Square.all()
    return new MoveFinder(this.currentSquare).legalMoves().filter(sq => !this.visited(sq))
  }

  visit(square) {
    if (!this.legalMovesFrom.some(sq => sq.equals(square))) throw new IllegalMoveError(`${square.notation} is not legal`)
    this.moves.push(square)
    this.redoStack = []
    return square
  }

  prev() { if (this.moves.length > 0) this.redoStack.push(this.moves.pop()) }
  next() { if (this.redoStack.length > 0) this.moves.push(this.redoStack.pop()) }
  toStart() { while (!this.atStart) this.prev() }
  toEnd() { while (!this.atEnd) this.next() }

  goTo(n) {
    const clamped = Math.max(0, Math.min(n, this.moves.length + this.redoStack.length))
    while (this.moves.length > clamped) this.prev()
    while (this.moves.length < clamped) this.next()
  }

  get atStart() { return this.moves.length === 0 }
  get atEnd() { return this.redoStack.length === 0 }

  get visitedCount() { return this.moves.length }
  get won() { return this.visitedCount === 64 }
  get stuck() { return this.visitedCount > 0 && !this.won && this.legalMovesFrom.length === 0 }

  notationPath() { return this.moves.map(sq => sq.notation) }
  fullNotationPath() { return [ ...this.moves, ...this.redoStack.slice().reverse() ].map(sq => sq.notation) }
}
